File: benchmarks/compare/raptor/node/src/raptor/QueueFactory.js
/**
 * Create a queue for the Raptor algorithm to use on each iteration of the algorithm.
 * All identifiers are integer indices (no string hashing in the hot path).
 */
export class QueueFactory {
  constructor(numStops, stopRoutes, stopRoutesBase, stopRoutesCount, routeStopPos) {
    this.stopRoutes = stopRoutes;
    this.stopRoutesBase = stopRoutesBase;
    this.stopRoutesCount = stopRoutesCount;
    this.routeStopPos = routeStopPos;
  }

  /**
   * Take the marked stops (integer indices) and return a Map<routeIdx, startPos>
   * where startPos is the earliest position in the route to start scanning from.
   */
  getQueue(markedStops) {
    const queue = new Map();
    const stopRoutes = this.stopRoutes;
    const stopRoutesBase = this.stopRoutesBase;
    const stopRoutesCount = this.stopRoutesCount;
    const routeStopPos = this.routeStopPos;

    for (const stopIdx of markedStops) {
      const base = stopRoutesBase[stopIdx];
      const count = stopRoutesCount[stopIdx];
      for (let i = 0; i < count; i++) {
        const routeIdx = stopRoutes[base + i];
        const pos = routeStopPos[routeIdx].get(stopIdx);
        if (pos === undefined) continue;
        const existing = queue.get(routeIdx);
        if (existing === undefined || pos < existing) {
          queue.set(routeIdx, pos);
        }
      }
    }

    return queue;
  }
}

File: benchmarks/compare/raptor/python/raptor/queue_factory.py
"""QueueFactory (mirrors src/raptor/QueueFactory.ts).

Builds a route->stop queue (insertion-ordered dict) from marked stops.
"""
from __future__ import annotations

from typing import Dict, List

from .gtfs import StopID

RouteID = str


class QueueFactory:
    def __init__(
        self,
        routes_at_stop: Dict[StopID, List[RouteID]],
        route_stop_index: Dict[RouteID, Dict[StopID, int]],
    ) -> None:
        self.routesAtStop = routes_at_stop
        self.routeStopIndex = route_stop_index

    def getQueue(self, marked_stops: List[StopID]) -> Dict[RouteID, StopID]:
        queue: Dict[RouteID, StopID] = {}

        for stop in marked_stops:
            for route_id in self.routesAtStop.get(stop, []):
                existing = queue.get(route_id)
                if existing is not None and self._is_stop_before(route_id, existing, stop):
                    queue[route_id] = existing
                else:
                    queue[route_id] = stop

        return queue

    def _is_stop_before(self, route_id: RouteID, stop_a: StopID, stop_b: StopID) -> bool:
        return self.routeStopIndex[route_id][stop_a] < self.routeStopIndex[route_id][stop_b]
